src/vm/runners/builtin_runner/bitwise.rs

```diff
@@ -58,9 +58,13 @@ impl BuiltinRunner for BitwiseBuiltinRunner {
         }
         let x_addr = MaybeRelocatable::from((address.segment_index, address.offset - index));
         let y_addr = x_addr.add_usize_mod(1, None);
-        if let (Ok(Some(MaybeRelocatable::Int(num_x))), Ok(Some(MaybeRelocatable::Int(num_y)))) =
-            (memory.get(&x_addr), memory.get(&y_addr))
-        {
+
+        let num_x = memory.get(&x_addr);
+        let num_y = memory.get(&y_addr);
+        if let (Ok(Some(MaybeRelocatable::Int(num_x))), Ok(Some(MaybeRelocatable::Int(num_y)))) = (
+            num_x.as_ref().map(|x| x.as_ref().map(|x| x.as_ref())),
+            num_y.as_ref().map(|x| x.as_ref().map(|x| x.as_ref())),
```

Comment:
These `map`s look weird. Why doesn't `as_ref` suffice?

Reply:
`num_x` and `num_y` are `Result<Option<Cow<MaybeRelocatable>>, _>`, while we're interested in getting a `&MaybeRelocatable`. `Result` and `Option` can be matched no problem, but to match the `Cow` we'd need two branches (`Owned` and `Borrowed`).

In order to avoid that, the first two `.as_ref().map(|_| todo!())` map the internal values of the `Result` (when `Ok`) and `Option` wrappers by reference (otherwise the value is moved out) while the last `x.as_ref()` converts `Cow<MaybeRelocatable>` to a `&MaybeRelocatable` regardless of its actual variant so that it can be matched without code duplication.

Using a single `.as_ref()` would provide a `Result<&Option<Cow<MaybeRelocatable>>, &_>` which does not remove the `Cow`.

A `match` could also be used, but it'd require `2^n` combined patterns (4 in this case), one for each combination of `Owned` and `Borrowed`. For example, something like this:
```rs
match (num_x, num_y) {
    (Ok(Some(Cow::Borrowed(num_x))), Ok(Some(Cow::Borrowed(num_y)))) |
    (Ok(Some(Cow::Borrowed(num_x))), Ok(Some(Cow::Owned(ref num_y)))) |
    (Ok(Some(Cow::Owned(ref num_x))), Ok(Some(Cow::Borrowed(num_y)))) |
    (Ok(Some(Cow::Owned(ref num_x))), Ok(Some(Cow::Owned(ref num_y)))) => {
        // `if` statement body here
        todo!()
    },
    _ => {},
}
```

Please let me know if a `match` would be preferred in this case over the `map` and `as_ref` chain.

Reply:
Now that I think about it, there's a potentially cleaner solution which requires a trait utility. An example can be found [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=213ed496e3fe6851e4b4fa4a8b09f949).

Basically, a trait is implemented for each `Result<Option<Cow<'a, T>>, E>` that provides a method to transform it to `Result<Option<&'a T>, E>`. It moves the `as_ref`-`map` chain into the trait while leaving a cleaner source code where used.

Reply:
> Please let me know if a match would be preferred in this case over the map and as_ref chain.

Most certainly not D:

> Basically, a trait is implemented for each Result<Option<Cow<'a, T>>, E> that provides a method to transform it to Result<Option<&'a T>, E>. It moves the as_ref-map chain into the trait while leaving a cleaner source code where used.

That does seem like a better solution in that it's less error prone and explodes the code much less. It's OK as it is given the explanation tho.